### DISCUSSION TITLE
Suggest a datasource integration 🔌

### DISCUSSION THREAD

**Discussion by melohagan** [Budibase/budibase#7807] [Ideas]
We are looking to greatly expand upon our available datasources!

What custom datasources would you most like to see? Comment below!

OG thread: https://github.com/Budibase/budibase/discussions/891

Current list of datasources can be found here: https://github.com/Budibase/plugins

**Comment by martkaczmarek** [comment]
I think Microsoft API called Graph could be a good idea. Through it you can access pretty much  any aspect of Microsoft 365 account and applications, including Sharepoint, Azure AD, Outlook etc.
This would be very helpful for those of us that are migrating from Microsoft Powerapps - many of the basic "blocks" in Powerapps are just Graph integrations.

**Reply by melohagan** [reply]
Good shout! We definitely need more Microsoft integrations!

**Reply by netsinerji** [reply]
I would add MS Excel (online) but I guess it falls under this one, being a Microsoft 365 app.

**Reply by martkaczmarek** [reply]
@netsinerji yup, it's all Graph

**Reply by acseven** [reply]
This is one is an absolutely must!

**Reply by andyburgessmd** [reply]
I agree on this one, would be really handy

**Comment by oldwired** [comment]
Supabase, but in the end it's just PostgreSQL with (a lot of useful) stuff added. But if we cloud use Supabase Auth, that would be nice.

**Reply by mshiner** [reply]
I second a 'proper' Supabase integration so we get all their sugar-coating (or is it added chilli.... not sure) on top of PostgreSQL

**Reply by melohagan** [reply]
I pushed up a Supabase plugin today if you want to give that a go? https://github.com/melohagan/budibase-datasource-supabase

I don't think it has the fully available functionality, but it has the core CRUD methods and filters.

**Reply by mshiner** [reply]
Great - I'll give that a go

**Reply by DV8821** [reply]
Cool!

**Comment by oldwired** [comment]
ODBC to have some kind of "generic Database interface", but through the way ODBC (drivers) work that could be difficult to achieve and manage.

**Comment by martkaczmarek** [comment]
As referenced in #7786 , it would be useful to have an Action in Automations to run another Automation - should be a quick job considering only Budibase's own API would be required.

**Comment by poirazis** [comment]
I am very favorably looking for a Data Source for [www.surrealdb.com ](https://surrealdb.com/) !

**Reply by melohagan** [reply]
Done! https://github.com/melohagan/budibase-datasource-surrealdb

**Comment by kmeesters** [comment]
May I suggest Directus (https://directus.io/). Although you can connect to the API or the backend DB directly, I think a direct data sources would be great. Also because that way i can also serve as authentication and RABC /ACL.
Moreover, I see Budibase and Directus as two very like-minded developments (same business model, same philosophy) just one focusing on the backend and one on the frontend.

**Comment by eftomi** [comment]
Oracle.

**Comment by gwrx** [comment]
I would love to connect to different Calendars as datasource. I.e. with CalDav as a basic implementation because there are caldav connectors to other syncprotocols like google calendar, microsoft calendar, iCal etc.

**Reply by vevedh** [reply]
Hello, @melohagan 

I try create a datasource https://github.com/Budibase/plugins/compare/main...vevedh:budibase-plugins:connector-ldap?expand=1
But I have this error :
@budibase/server:     _events: [Object: null prototype] {
@budibase/server:       error: [Function (anonymous)],
@budibase/server:       message: [Function: bound _handleSendResponse]
@budibase/server:     },
@budibase/server:     _eventsCount: 2,
@budibase/server:     _maxListeners: undefined,
@budibase/server:     [Symbol(kCapture)]: false
@budibase/server:   },
@budibase/server:   messageDetailsByMessageId: {},
@budibase/server:   socketDataHandler: [Function (anonymous)],
@budibase/server:   clientOptions: {
@budibase/server:     url: 'ldap://agglo.local',
@budibase/server:     timeout: 0,
@budibase/server:     connectTimeout: 0,
@budibase/server:     strictDN: true
@budibase/server:   },
@budibase/server:   secure: false,
@budibase/server:   host: 'agglo.local',
@budibase/server:   port: 389
@budibase/server: }
@budibase/server: ERROR [1664407457458] (29408 on svrapps.agglo.local): AD Ldap error: Error: crypto.getRandomValues() not supported. See https://github.com/uuidjs/uuid#getrandomvalues-not-supported
@budibase/server:     Error: AD Ldap error: Error: crypto.getRandomValues() not supported. See https://github.com/uuidjs/uuid#getrandomvalues-not-supported
@budibase/server:         at integration.<anonymous> (/tmp/.budibase/datasource/budibase-datasource-ldap:1:97912)
@budibase/server:         at Generator.throw (<anonymous>)
@budibase/server:         at a (/tmp/.budibase/datasource/budibase-datasource-ldap:1:454)
@budibase/server:         at runMicrotasks (<anonymous>)
@budibase/server:         at processTicksAndRejections (node:internal/process/task_queues:96:5)
@budibase/server: Trace: Error: AD Ldap error: Error: crypto.getRandomValues() not supported. See https://github.com/uuidjs/uuid#getrandomvalues-not-supported
@budibase/server:     at integration.<anonymous> (/tmp/.budibase/datasource/budibase-datasource-ldap:1:97912)
@budibase/server:     at Generator.throw (<anonymous>)
@budibase/server:     at a (/tmp/.budibase/datasource/budibase-datasource-ldap:1:454)
@budibase/server:     at runMicrotasks (<anonymous>)
@budibase/server:     at processTicksAndRejections (node:internal/process/task_queues:96:5) {
@budibase/server:   expose: true,
@budibase/server:   statusCode: 400,
@budibase/server:   status: 400
@budibase/server: }
@budibase/server:     at /home/budibase/packages/server/src/api/index.js:78:15
@budibase/server:     at Generator.throw (<anonymous>)
@budibase/server:     at rejected (/home/budibase/packages/server/src/api/index.js:6:65)
@budibase/server:     at runMicrotasks (<anonymous>)
@budibase/server:     at processTicksAndRejections (node:internal/process/task_queues:96:5)@budibase/server: bb-alert: Uncaught exception. : {"stack":"Error: read ECONNRESET\n    at TCP.onStreamRead (node:internal/stream_base_commons:217:20)\n    at TCP.callbackTrampoline (node:internal/async_hooks:130:17)","message":"read ECONNRESET","errno":-104,"code":"ECONNRESET","syscall":"read"}

Can you help me ?

**Reply by melohagan** [reply]
Looks like you're missing a library/import just: https://github.com/uuidjs/uuid/issues/514

**Comment by acseven** [comment]
SOAP API

**Reply by melohagan** [reply]
Getting flash backs of the horror of having to use SOAP that one time 🤕

**Reply by acseven** [reply]
Unfortunately there are some services that only make data available via SOAP API...! :-/

**Reply by acseven** [reply]
@melohagan can you say if a SOAP data source integration is likely to happen in the future?

**Reply by melohagan** [reply]
Hey @acseven I have built an extremely rudimentary SOAP connector here: https://github.com/melohagan/budibase-datasource-soap

The reason I didn't flesh it out further is because there's a number of different directions you could go with it to try and make it a bit nicer to use.
For example the response is just a stringified xml envelope. How would you best want to handle that? 
I could add a field that allows you to specify the path to the response. 

E.g. if a response is:
```
<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
    <soap:Body>
        <m:NumberToWordsResponse xmlns:m="http://www.dataaccess.com/webservicesserver/">
            <m:NumberToWordsResult>five hundred </m:NumberToWordsResult>
        </m:NumberToWordsResponse>
    </soap:Body>
</soap:Envelope>
```

Then maybe you could provide a field with the path: `NumberToWordsResponse.NumberToWordsResult`
which then the connector would return `five hundred`
Or would you want the response to be a JSON object like:

```
{
  'NumberToWordsResponse': {
    'NumberToWordsResult': 'five hundred '
  }
}
```

Feel free to raise any ideas/issues on that GitHub repository!

**Comment by NicolasFabre81** [comment]
Our new data source portal standard is **DENODO**, so adding Denodo would be great for us.

**Comment by TheGameVIX** [comment]
If possible I personally would really like to see a smartsheets integration.

**Comment by mackrp** [comment]
I’d like to see integration with iAuditor (Safetyculture)

**Comment by jensjohansen** [comment]
I suggest adding Trino as a data source. Trino would give Budibase access to a large number of data lake technologies for real-time analytics, such as Iceberg stored on S3, Cassandra, Druid, Pinot, and Pravega.

**Reply by jensjohansen** [reply]
https://trino.io/

**Comment by 5hanth** [comment]
would anyone be interested in a datasource for [Nostr](https://github.com/nbd-wtf/nostr-tools). It relies on websockets.

**Comment by harin** [comment]
I would like MS SQL with the option to trust the server certificate.

**Comment by rbq** [comment]
Mapping to/from Kubernetes resources would be quite useful. I'm thinking e.g. Crossplane stuff …

**Comment by kerbersystems** [comment]
Azure Data Lake (Gen2) integration similar to the current S3 integration. I know of multiple billion dollar orgs that would love to adopt BB but have no way to send files to ALDS Gen2.

**Comment by gorkemcetin** [comment]
Retable is an Airtable alternative with a focus on simplicity, user experience and the ability to self-host (albeit not open source). Our customers have been asking for a Budibase integration and I kindly ask you to add Retable integration. 

Web page: https://retable.io
API documentation: https://docs.retable.io/retable-user-guide/retable-api/api

Thank you for your consideration!

**Comment by ghost** [comment]
@melohagan Can you tell me how budibase connects to supabase locally
